### PR TITLE
#102 Automated unit test case written successfully for kill_child_process.py

### PR DESCRIPTION
**Pull request by SinghHimadri**
#102 
Closes: # (issue number)

## Proposed changes
Successful test added for kill_chil_process
Brief description of what is fixed or changed
To create a child process subprocess library has been used a function get_test_subprocess() which does nothing for 30 sec has been written. This test works in all MACOS, WINDOWS, and Linux. Since this test deals with os part all care has been taken and os.fork had many limitations so subprocess is used. The logic for the test is a child is created its process id is stored then this child process is killed using kill_child_process function and then checks is this id still exists in our os using psutil.pid_exists(test_pid) we self assert false. As the process is killed so it returns false and we assert it.
## Types of changes
A test folder has been added. This folder contains test_kill_child_processes and test_func which contains a function that creates child process which does nothing for the next 30 sec.
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (Documentation content changed)
- [ ] Other (please describe): 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Screenshots
I have written the test that works successfully and the function passes the test i.e. used kill_child_process function used. I am sharing results.Attaching screenshots
Screenshot from 2020-03-30 15-39-10

Please attach the screenshots of the changes made in case of a change in user interface

## Other information
I would request to not unassign me or close this issue since this is the first automated unit test added. I am writing more tests and will be sending their PR soon. Also, I haven't updated documentation because more tests are to be added. I'll update once all test cases are added.
Any other information that is important to this pull request

### CODE REVIEW

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 22
           

Complexity increasing per file
==============================
- wplay/tests/test_kill_child_process.py  1
- wplay/tests/test_func.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955&bid=17247658)

File: wplay/tests/test_kill_child_process.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        p.wait()
+        self.assertFalse(psutil.pid_exists(test_pid))
+
+if __name__ == '__main__':
+	unittest.main()
+'''
+if __name__ == '__main__':
+    from psutil.tests.runner import run
+    run(__file__)
+'''
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+from psutil import POSIX
+
+from wplay.tests import test_func
+from wplay.utils import kill_process
+from wplay.tests import get_test_subprocess
+from wplay.tests import PYTHON_EXE
+from psutil.tests import retry_on_failure
+from psutil.tests import safe_rmpath
+from wplay.tests import TESTFILE_PREFIX
+from wplay.tests import TESTFN
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused TESTFN imported from wplay.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            attempt(sys.executable) or \
+            attempt(os.path.realpath(sys.executable)) or \
+            attempt(which("python%s.%s" % sys.version_info[:2])) or \
+            attempt(psutil.Process().exe())
+        if not exe:
+            raise ValueError("can't find python exe real abspath")
+        return exe
+    else:
+        exe = os.path.realpath(sys.executable)
+        assert os.path.exists(exe), exe
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+TESTFN = os.path.join(os.path.realpath(os.getcwd()), TESTFILE_PREFIX)
+TESTFN = TESTFN + str(os.getpid())
+
+_TESTFN = TESTFN + '-internal'
+
+
+def _get_py_exe():
+    def attempt(exe):
+        try:
+            subprocess.check_call(
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import psutil 
+
+from psutil import POSIX
+
+from wplay.tests import test_func
+from wplay.utils import kill_process
+from wplay.tests import get_test_subprocess
+from wplay.tests import PYTHON_EXE
+from psutil.tests import retry_on_failure
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused retry_on_failure imported from psutil.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
+#region imports
+import os
+import signal
+import subprocess
+import sys
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import sys](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import time
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import time](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
+#region imports
+import os
+import signal
+import subprocess
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+from psutil import POSIX
+
+from wplay.tests import test_func
+from wplay.utils import kill_process
+from wplay.tests import get_test_subprocess
+from wplay.tests import PYTHON_EXE
+from psutil.tests import retry_on_failure
+from psutil.tests import safe_rmpath
+from wplay.tests import TESTFILE_PREFIX
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused TESTFILE_PREFIX imported from wplay.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
+#region imports
+import os
+import signal
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import signal](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import subprocess
+import sys
+import time
+import unittest
+
+import psutil 
+
+from psutil import POSIX
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused POSIX imported from psutil](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 12]
```diff
@@ -0,0 +1,157 @@
+#region imports
+import functools
+import os
+import shutil
+import stat
+import subprocess
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py (hunk 8 above)
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import subprocess](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+from wplay.tests import test_func
+from wplay.utils import kill_process
+from wplay.tests import get_test_subprocess
+from wplay.tests import PYTHON_EXE
+from psutil.tests import retry_on_failure
+from psutil.tests import safe_rmpath
+from wplay.tests import TESTFILE_PREFIX
+from wplay.tests import TESTFN
+from wplay.tests import unittest
+from wplay.tests import wait_for_pid
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused wait_for_pid imported from wplay.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        # give it some more time to allow better initialization
+        time.sleep(0.01)
+
+
+def wait_for_file(fname, delete=True, empty=False):
+    """Wait for a file to be written on disk with some content."""
+    with open(fname, "rb") as f:
+        data = f.read()
+    if not empty:
+        assert data
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+#endregion
+
+
+__all__ = [
+    # constants
+    'DEVNULL' , 'PYTHON_EXE', 'TESTFILE_PREFIX' , 'TESTFN',
+    # subprocesses
+    'get_test_subprocess',
+    # test utils
+    'unittest' ,
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Undefined variable name 'unittest' in __all__](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import time
+import unittest
+
+import psutil 
+
+from psutil import POSIX
+
+from wplay.tests import test_func
+from wplay.utils import kill_process
+from wplay.tests import get_test_subprocess
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused get_test_subprocess imported from wplay.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+from wplay.tests import PYTHON_EXE
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused PYTHON_EXE imported from wplay.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+from psutil.tests import retry_on_failure
+from psutil.tests import safe_rmpath
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused safe_rmpath imported from psutil.tests](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
+#region imports
+import os
+import signal
+import subprocess
+import sys
+import time
+import unittest
+
+import psutil
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 20]
```diff
@@ -0,0 +1,44 @@
+#region imports
+import os
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import os](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+        safe_rmpath(_TESTFN)
+        pyline = "from time import sleep;" \
+                 "open(r'%s', 'w').close();" \
+                 "sleep(30);" % _TESTFN
+        cmd = [PYTHON_EXE, "-c", pyline]
+        sproc = subprocess.Popen(cmd, **kwds)
+        _subprocesses_started.add(sproc)
+        #wait_for_file(_TESTFN, delete=True, empty=True)
+    else:
+        sproc = subprocess.Popen(cmd, **kwds)
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- wplay/tests/test_kill_child_process.py  1
- wplay/tests/test_func.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955&bid=17247658)

File: wplay/tests/test_func.py [hunk 22]
```diff
@@ -0,0 +1,155 @@
+#region imports
+import functools
+import os
+import shutil
+import stat
+import subprocess
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+
+
+def wait_for_file(fname, delete=True, empty=False):
+    """Wait for a file to be written on disk with some content."""
+    with open(fname, "rb") as f:
+        data = f.read()
+    if not empty:
+        assert data
+    if delete:
+        safe_rmpath(fname)
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+import signal
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import signal](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+                 "open(r'%s', 'w').close();" \
+                 "sleep(30);" % _TESTFN
+        cmd = [PYTHON_EXE, "-c", pyline]
+        sproc = subprocess.Popen(cmd, **kwds)
+        _subprocesses_started.add(sproc)
+        #wait_for_file(_TESTFN, delete=True, empty=True)
+    else:
+        sproc = subprocess.Popen(cmd, **kwds)
+        _subprocesses_started.add(sproc)
+        wait_for_pid(sproc.pid)
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+import subprocess
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import subprocess](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 27]
```diff
@@ -0,0 +1,25 @@
+#region imports
+import os
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Unused import os](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py (hunk 26 above)
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_kill_child_process.py [hunk 28]
```diff
@@ -0,0 +1,25 @@
...
+import signal
+import subprocess
+import unittest
+import psutil
```
Comment (rpotter12) [inline wplay/tests/test_kill_child_process.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+
+_TESTFN = TESTFN + '-internal'
+
+
+def _get_py_exe():
+    def attempt(exe):
+        try:
+            subprocess.check_call(
+                [exe, "-V"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        except Exception:
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+            attempt(which("python%s.%s" % sys.version_info[:2])) or \
+            attempt(psutil.Process().exe())
+        if not exe:
+            raise ValueError("can't find python exe real abspath")
+        return exe
+    else:
+        exe = os.path.realpath(sys.executable)
+        assert os.path.exists(exe), exe
+        return exe
+
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- wplay/tests/test_kill_child_process.py  1
- wplay/tests/test_func.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955&bid=17247658)

File: wplay/tests/test_func.py [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+        safe_rmpath(_TESTFN)
+        pyline = "from time import sleep;" \
+                 "open(r'%s', 'w').close();" \
+                 "sleep(30);" % _TESTFN
+        cmd = [PYTHON_EXE, "-c", pyline]
+        sproc = subprocess.Popen(cmd, **kwds)
+        _subprocesses_started.add(sproc)
+        #wait_for_file(_TESTFN, delete=True, empty=True)
+    else:
+        sproc = subprocess.Popen(cmd, **kwds)
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+TESTFN = os.path.join(os.path.realpath(os.getcwd()), TESTFILE_PREFIX)
+TESTFN = TESTFN + str(os.getpid())
+
+_TESTFN = TESTFN + '-internal'
+
+
+def _get_py_exe():
+    def attempt(exe):
+        try:
+            subprocess.check_call(
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 33]
```diff
@@ -0,0 +1,155 @@
...
+            attempt(sys.executable) or \
+            attempt(os.path.realpath(sys.executable)) or \
+            attempt(which("python%s.%s" % sys.version_info[:2])) or \
+            attempt(psutil.Process().exe())
+        if not exe:
+            raise ValueError("can't find python exe real abspath")
+        return exe
+    else:
+        exe = os.path.realpath(sys.executable)
+        assert os.path.exists(exe), exe
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py [hunk 34]
```diff
@@ -0,0 +1,155 @@
...
+        # give it some more time to allow better initialization
+        time.sleep(0.01)
+
+
+def wait_for_file(fname, delete=True, empty=False):
+    """Wait for a file to be written on disk with some content."""
+    with open(fname, "rb") as f:
+        data = f.read()
+    if not empty:
+        assert data
```
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: wplay/tests/test_func.py (hunk 22 above)
Comment (rpotter12) [inline wplay/tests/test_func.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- test_kill_child_process.py  1
- test_func.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955&bid=17247658)

File: test_func.py [hunk 35]
```diff
@@ -0,0 +1,155 @@
...
+        safe_rmpath(_TESTFN)
+        pyline = "from time import sleep;" \
+                 "open(r'%s', 'w').close();" \
+                 "sleep(30);" % _TESTFN
+        cmd = [PYTHON_EXE, "-c", pyline]
+        sproc = subprocess.Popen(cmd, **kwds)
+        _subprocesses_started.add(sproc)
+        #wait_for_file(_TESTFN, delete=True, empty=True)
+    else:
+        sproc = subprocess.Popen(cmd, **kwds)
```
Comment (rpotter12) [inline test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: test_func.py [hunk 36]
```diff
@@ -0,0 +1,155 @@
+#region imports
+import functools
+import os
+import shutil
+import stat
+import subprocess
```
Comment (rpotter12) [inline test_func.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: test_func.py [hunk 37]
```diff
@@ -0,0 +1,155 @@
...
+            attempt(sys.executable) or \
+            attempt(os.path.realpath(sys.executable)) or \
+            attempt(which("python%s.%s" % sys.version_info[:2])) or \
+            attempt(psutil.Process().exe())
+        if not exe:
+            raise ValueError("can't find python exe real abspath")
+        return exe
+    else:
+        exe = os.path.realpath(sys.executable)
+        assert os.path.exists(exe), exe
```
Comment (rpotter12) [inline test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: test_func.py [hunk 38]
```diff
@@ -0,0 +1,155 @@
...
+TESTFN = os.path.join(os.path.realpath(os.getcwd()), TESTFILE_PREFIX)
+TESTFN = TESTFN + str(os.getpid())
+
+_TESTFN = TESTFN + '-internal'
+
+
+def _get_py_exe():
+    def attempt(exe):
+        try:
+            subprocess.check_call(
```
Comment (rpotter12) [inline test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: test_func.py [hunk 39]
```diff
@@ -0,0 +1,155 @@
...
+        # give it some more time to allow better initialization
+        time.sleep(0.01)
+
+
+def wait_for_file(fname, delete=True, empty=False):
+    """Wait for a file to be written on disk with some content."""
+    with open(fname, "rb") as f:
+        data = f.read()
+    if not empty:
+        assert data
```
Comment (rpotter12) [inline test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

**Comment by SinghHimadri** [comment]
The test is running successfully I don't think there are any issues. Please review @rpotter12 @xandao6

**Comment by priyansh19** [comment]
Hi @SinghHimadri 
Please check your Pr there is increased complexity with your files 
- test_kill_child_process.py  by 1
- test_func.py  by 7

Do resolve them and update your PR :+1:

**Comment by xandao6** [comment]
Changed to medium difficulty because just solve one of the many possible tests.

**Comment by priyansh19** [comment]
@SinghHimadri add them in the test folder and work on further tests :+1: 
@xandao6 She will be adding more test cases in the future.

**Comment by rpotter12** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- tests/test_func.py  7
- tests/test_kill_child_process.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955&bid=17247658)

File: tests/test_func.py [hunk 40]
```diff
@@ -0,0 +1,155 @@
...
+TESTFN = os.path.join(os.path.realpath(os.getcwd()), TESTFILE_PREFIX)
+TESTFN = TESTFN + str(os.getpid())
+
+_TESTFN = TESTFN + '-internal'
+
+
+def _get_py_exe():
+    def attempt(exe):
+        try:
+            subprocess.check_call(
```
Comment (rpotter12) [inline tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: tests/test_func.py [hunk 41]
```diff
@@ -0,0 +1,155 @@
...
+        safe_rmpath(_TESTFN)
+        pyline = "from time import sleep;" \
+                 "open(r'%s', 'w').close();" \
+                 "sleep(30);" % _TESTFN
+        cmd = [PYTHON_EXE, "-c", pyline]
+        sproc = subprocess.Popen(cmd, **kwds)
+        _subprocesses_started.add(sproc)
+        #wait_for_file(_TESTFN, delete=True, empty=True)
+    else:
+        sproc = subprocess.Popen(cmd, **kwds)
```
Comment (rpotter12) [inline tests/test_func.py]: Codacy Issue found: [subprocess call - check for execution of untrusted input.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: tests/test_func.py [hunk 42]
```diff
@@ -0,0 +1,155 @@
+#region imports
+import functools
+import os
+import shutil
+import stat
+import subprocess
```
Comment (rpotter12) [inline tests/test_func.py]: Codacy Issue found: [Consider possible security implications associated with subprocess module.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: tests/test_func.py [hunk 43]
```diff
@@ -0,0 +1,155 @@
...
+        # give it some more time to allow better initialization
+        time.sleep(0.01)
+
+
+def wait_for_file(fname, delete=True, empty=False):
+    """Wait for a file to be written on disk with some content."""
+    with open(fname, "rb") as f:
+        data = f.read()
+    if not empty:
+        assert data
```
Comment (rpotter12) [inline tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

File: tests/test_func.py [hunk 44]
```diff
@@ -0,0 +1,155 @@
...
+            attempt(sys.executable) or \
+            attempt(os.path.realpath(sys.executable)) or \
+            attempt(which("python%s.%s" % sys.version_info[:2])) or \
+            attempt(psutil.Process().exe())
+        if not exe:
+            raise ValueError("can't find python exe real abspath")
+        return exe
+    else:
+        exe = os.path.realpath(sys.executable)
+        assert os.path.exists(exe), exe
```
Comment (rpotter12) [inline tests/test_func.py]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/rpotter12/whatsapp-play/pullRequest?prid=5264955)

**Comment by SinghHimadri** [comment]
> LGTM.....I have not tested it yet.

I have added command for running tests in test_kill_chil_process. You can test it 🙂

**Pull request merged**